Packages/ClientRuntime/Sources/Networking/Http/SdkHttpRequest.swift

```diff
@@ -29,12 +29,22 @@ public class SdkHttpRequest {
     }
 }
 
+// Create a `CharacterSet` of the characters that need not be percent encoded in the
+// resulting URL.  This set consists of alphanumerics plus underscore, dash, tilde, and
+// period.  Any other character should be percent-encoded when used in a path segment.
+// Forward-slash is added as well because the segments have already been joined into a path.
+//
+// See, for URL-allowed characters:
+// https://www.rfc-editor.org/rfc/rfc3986#section-2.3
+private let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "/_-.~"))
```

Comment:
does S3 follow the same spec?

It would be nice to add a reference to the S3 documentation.

Reply:
As far as I can tell, S3 docs never explicitly say how to encode they key for use in the URI.  But it makes sense that it must be at least encoded per the web standards to form a valid URI path.

S3 has [this doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html) that discusses characters that are & aren't allowed in keys, but it doesn't really talk about how to encode them in a URL.

Last, we must remember that this is in Smithy and that this code is not specific to AWS or S3.  So this must behave for any web service.  Encoding all but the explicitly allowed characters should be the most safe option.